### ISSUE TITLE
Reserved characters in file system object names

### ISSUE THREAD

**Issue by marcastel** [sveltejs/kit#894]
**Describe the bug**
Any file system object whose name contains UNIX reserved characters, and most importantly the dollar sign (<kbd>$</kbd>), is clearly against POSIX best practices, and qualifies, IMHO, as a bug.

**Expected behavior**
Comply to operating system best practices when naming file system objects.

**Notes**
I am discovering SvelteKit... and this looks like a great piece of software. But using the dollar sign in file names really bugs me. Sorry.

Of all ASCII characters why choose a symbol that carries a very special meaning for all UNIX-flavoured operating systems. Other symbols are conventionally used to distinguish special file system objects from others (e.g. <kbd>.</kbd>, <kbd>_</kbd>, <kbd>~</kbd>, <kbd>%</kbd>, or <kbd>!</kbd>). Why not use one of these?

Most systems are UTF8 today, so one could even use such extended symbols (e.g. <kbd>ƶ</kbd> → `ƶlayout.svelte`)
-- though I would not recommend this.

I strongly encourage to strip/replace this dollar sign from all file names, ASAP.

**Comment by Rich-Harris** [comment]
'Best practices' is handwavy — can you elaborate on why `$` is a problem?

**Comment by Conduitry** [comment]
There were some discussions that I recall on Sapper where people were having specific problems with filenames containing bad characters on some hosting providers. I've found https://github.com/sveltejs/sapper/issues/747 just now but I don't remember whether there were others.

**Comment by rmunn** [comment]
There are only two reserved characters in UNIX filenames, `/` and `\0` (the null character). All others are fair game, including `$`. Space characters in filenames actually cause more problems for shell scripts than dollar signs, because once a dollar sign has made it into the parameters it won't be shell-expanded, whereas spaces cause problems if your Bash script is written with `$@` instead of `"$@"`. The only thing you need to do is quote with single-quotes instead of double-quotes: `"$layout"` will be expanded to the value of the `layout` environment variable (or a blank string if there is none), whereas `'$layout'` will be the literal string `$layout`.

Speaking as someone who has used Linux since 1999, the dollar sign doesn't bug me at all. But it's worth considering whether to change to a different character. Once Svelte-Kit is officially released it will be far too late, but right now the `@` character should be considered. It already has a sort-of de-facto usage in Javascript, and `@lib` as an alias for `src/lib` would make a lot of sense, etc.

But if `$` is going to be replaced, do NOT use the tilde (`~`). It will cause just as many problems, or maybe even more, than `$` would.

**Comment by Conduitry** [comment]
Without more details about why `$` is a problem besides 'best practices', I'm going to close this issue.

**Comment by marcastel** [comment]
> 'Best practices' is handwavy — can you elaborate on why `$` is a problem?

Sorry for late reply.  OOO. I see that you have already adopted changes. Thanks for considering this  :-)

For the record. [David Wheeler](https://dwheeler.com/essays/fixing-unix-linux-filenames.html) elaborates on this. <kbd>$</kbd> is a meta character as are <kbd>*</kbd>, <kbd>?</kbd>, <kbd>:</kbd>, <kbd>[</kbd>, <kbd>]</kbd>, <kbd>"</kbd>, <kbd><</kbd>, <kbd>></kbd>, <kbd>|</kbd>, <kbd>(</kbd>, <kbd>)</kbd>, <kbd>{</kbd>, <kbd>}</kbd>, <kbd>&</kbd>, <kbd>'</kbd>, <kbd>!</kbd>, <kbd>\\</kbd>, and <kbd>;</kbd><sup>(b)</sup>. The C language primitives that handle *exec()'d* or *shell*'d<sup>(a)</sup> commands will treat them specifically unless they are properly escaped. As for any programming language, using meta-characters should be done with caution. Very much like when using <kbd>&lt;</kbd> or <kbd>&gt;</kbd> in SGML-derived markup languages such as HTML. Basing a naming convention on the use of such meta-characters is, IMHO, a **bad practice**.

Further allowing meta-characters (and the dollar sign in particular) can lead to unsuspected exploits. The dollar sign orders the underlying system to **expand** or **execute** something. `$USER` will reveal your login name. `${.sh.*}` will reveal details of your shell interpreter. `$(rm -rf /)` will wipe your disk. In the early days of the World Wide Web and the first CGI scripts, this was a common security issue. Exploits via query strings is still a major security concern today. I agree that, by design, Svelte will never expose this directly in production. Nonetheless adopting _conventional_ best practices here would be my recommendation.

David Wheeler's quote says it all:
> Seek freedom and become captive of your desires, seek discipline and find your liberty.

<sup>(a)</sup> Extra risks may arise with different handling by the platform's specific C library implementation, or by additional functionality introduced by non-POSIX compliant shells, Bash, Fish, etc. The `$@` vs `"$@"` debate fails in most scenarios beyond direct execution of simple shells because of the **backslash hell** to properly escape a string which will be transmitted by multiple programs (e.g. ssh connections or GitHub actions).

<sup>(b)</sup> Note that I have NOT included the <kbd>~</kbd> character. It is a meta character, ONLY for shells and ONLY if it is used as the leading character followed by a <kbd>/</kbd>; in such cases it represents a users' HOME directory. This character has commonly been used as a leading/trailing symbol to mark temporary files and is not subject to particular handling by C library functions.

BTW. Bash is GNU. Bash is not POSIX. Nor is Linux. Though they strive to be more and more. Younger generations tend to forget the very hard time we had in the 70s and 80s with the non-standardisation of UNIX. [POSIX](https://en.wikipedia.org/wiki/POSIX) is what unified the myriad versions out there. Today's UNIXes are all POSIX-compliant: AIX, HPUX, ... even macOS. POSIX should be foundational to any best practices and conventions. So much so,  that POSIX is now an international standard (ISO 9945).

**Comment by Radioz028** [comment]
I agree with @marcastel and I think this issue should be re-opened.

I have just experienced a problem, first-hand, caused by the use of these special characters in file names. Specifically, when using the [Sublime SFTP plugin](https://codexns.io/products/sftp_for_sublime) to edit a SvelteKit project on a remote host, I was unable to read files within parameter directories, e.g. `./[productId]/+page.svelte`. It just can't seem to comprehend directories with square brackets. This is a popular plugin, and not an uncommon use-case.

Now, Is this an internal problem with the Sublime SFTP Plugin? Absolutely. It's their fault for not handling special characters properly.

However, on a practical level, everybody knows that including special characters in file names can cause random issues like this, because everybody knows that software is imperfect. As technical people, we are aware of what things are more likely to cause issues, so we choose safer options to minimise the surface for bugs.

Take a look at the SvelteKit source code (or most other projects). File names are broadly alphanumeric, lowercase, don't contain spaces. As developers we've settled on these conventions, in part, because they cause the least hassle. Metacharacters would need escaping, mixed case behaves differently on different filesystems, spaces need extra thought in shell scripting. At large we've settled on the idea that special characters in file names are a bad idea, not because there's a hard standard against them, but because they're more likely to be problematic and we'd rather not deal with the headache.

There's the purist mentality that anything except nulls and backslashes is fair game. Then there's practical mentality that, in a popular piece of software to be many in wildly different environments, picking the safer options decreases the surface area for bugs and decreases the overall amount of time people spend on silly issues.

Also, from a security perspective, as @marcastel also notes, these things get mishandled _all the time_. It's not SvelteKit that mishandles them, and it's not SvelteKit's fault that other software mishandles them, but SvelteKit would be playing a part in encouraging the use of "riskier" characters.

That is my opinion. Special characters look nice, but have higher potential for problems. There are plenty of perfectly fine naming conventions that don't involve special characters.
